### PR TITLE
[WIP] - [Metrics ServerApp] - New metrics serverapp

### PR DESCRIPTION
**Pull request by Jklawreszuk**
# PR Details

## Description

PR migrates current ASP.NET project to ASP.NET Core + EntityFramework/Dapper. I've also rearranged code refactored code to make it more readable. Additionally, project has Swagger support to easily call endpoints.

## Types of changes

- [x] Docs change / refactoring / dependency upgrade
- [x] Bug fix (non-breaking change which fixes an issue)

## Todo

- [ ] - Finish following endpoints : 
    - [ ] `ActivityMetrics.Controller. GetCrashesPerVersion` - takes about 2 minutes to load. I was not able to materialize the query. Probably need to create a `MetricCache` table, create a job using i.e HangFire lib or SQL agent  
    - [ ] `ActivityMetrics.Controller.GetHighUsage` - I need to test this, sometimes it takes too long to load.
- [ ] - Create SQL script to apply EF Core in production
- [ ] - Complete JS client
- [ ] - Implement database seeder
- [ ] - Test app when it is finished

I am leaving this PR as a WIP because I need a break. Though if you want to help, please feel free to collaborate and contribute your ideas or suggestions through my fork or anywhere😊

### CODE REVIEW

**Comment by VaclavElias** [comment]
@Jklawreszuk 

Could you send met a link / zip to the database dump you used and I can test it?

File: sources/metrics/Stride.Metrics.ServerApp/Controllers/Api/ActivityMetricsController.cs [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+    /// <summary>
+    /// Gets the usage count per version in the last 30 days.
+    /// </summary>
+    /// <returns>The usage count per version in the last 30 days.</returns>
+
+    [HttpGet("projects-users")]
+    public IEnumerable<ProjectsUsersAggregation> GetProjectsUsers()
+    {
+        var result = _metricDbContext.MetricEvents
+            .Where(a => a.MetricId == 19)
```
Comment (manio143) [inline sources/metrics/Stride.Metrics.ServerApp/Controllers/Api/ActivityMetricsController.cs:47]: Can you put that 19 in a names constant and explain which metric it represents (like `const int XxxMetricId = 19;`)?

File: sources/metrics/Stride.Metrics.ServerApp/Controllers/Api/ActivityMetricsController.cs [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+            .GroupBy(m => new { m.Month, m.Year })
+            .Select(g => new ProjectsUsersAggregation
+            (
+                g.Key.Year,
+                g.Key.Month,
+                g.Sum(x => x.Count > 1 ? 1 : 0),
+                g.Sum(x => x.Count > 3 ? 1 : 0),
+                g.Sum(x => x.Count > 5 ? 1 : 0)
+            ))
+            .ToList();
```
Comment (manio143) [inline sources/metrics/Stride.Metrics.ServerApp/Controllers/Api/ActivityMetricsController.cs:77]: nit: `ToListAsync()` like below - applies in other methods as well

File: sources/metrics/Stride.Metrics.ServerApp/Controllers/Api/InstallationMetricsController.cs [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+
+        return installsPerMonth;
+    }
+
+    /// <summary>
+    /// Gets the installs count in the last 30 days per day.
+    /// </summary>
+    /// <returns>The installs count in the last 30 days per day</returns>
+    [HttpGet]
+    [Route("installs-last-days")]
```
Comment (manio143) [inline sources/metrics/Stride.Metrics.ServerApp/Controllers/Api/InstallationMetricsController.cs:103]: nit: route path can be moved up into the `[HttpGet]`

File: sources/metrics/Stride.Metrics.ServerApp/Controllers/Api/MetricApiController.cs [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+        // If special guid, then don't store anything
+        var ipAddress = _httpContextAccessor.GetIPAddress();
+
+        if (newMetric.SpecialId == ByPassSpecialGuid) //filter out SSKK by default
+        {
+            Trace.TraceInformation("/api/push-metric (origin: {0}) with SpecialGuid not saved {1}", ipAddress, JsonConvert.SerializeObject(newMetric));
+            return Ok();
+        }
+
+        var clock = Stopwatch.StartNew();
```
Comment (manio143) [inline sources/metrics/Stride.Metrics.ServerApp/Controllers/Api/MetricApiController.cs:51]: Do we need to stop the Stopwatch before exiting this method?

File: sources/metrics/Stride.Metrics.ServerApp/Extensions/MetricDbContextExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+            InstallId = installId,
+            SessionId = newMetric.SessionId,
+            MetricId = metricId,
+            IPAddress = ipAddress,
+            EventId = newMetric.EventId,
+            MetricValue = newMetric.Value ?? string.Empty,
+            Timestamp = overrideTime ?? default
+        };
+
+        var metricAsJson = JsonConvert.SerializeObject(newMetric);
```
Comment (manio143) [inline sources/metrics/Stride.Metrics.ServerApp/Extensions/MetricDbContextExtensions.cs:101]: Would it make sense to move this serialization into the if block - this way we don't do it unnecessarily on the happy path.

File: sources/metrics/Stride.Metrics.ServerApp/Program.cs [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+using Stride.Metrics.ServerApp.Data;
+using Stride.Metrics.ServerApp.Extensions;
+
+var builder = WebApplication.CreateBuilder(args);
+
+// Add SQL Server database 
+builder.Services.AddDbContext<MetricDbContext>(opt =>
+{
+    opt.UseSqlServer(builder.Configuration.GetConnectionString("DefaultConnection"),
+    options => options.CommandTimeout(999999999));
```
Comment (manio143) [inline sources/metrics/Stride.Metrics.ServerApp/Program.cs:12]: I would strongly suggest making this smaller 😄
15 minutes timeout should probably be enough - it's not like people will be waiting that long before getting discouraged anyways and maybe the query needs to be further optimized or db upgraded if it takes that long...

File: sources/metrics/Stride.Metrics.ServerApp/Readme.md [hunk 7]
```diff
@@ -0,0 +1,16 @@
+# Stride Metrics ServerApp
+
+## Prerequisites
+
+- `ASPNET_ENVIRONMENT`: Set this variable to `Development` if you want to run the application in a testing environment. In the development environment, the application will connect to a non-Local SQL Server, allowing you to run the application on any operating system (i.e. [by using docker image](https://hub.docker.com/_/microsoft-mssql-server).).
+- `--seed-metrics` (recommended): Set this argument if you want to seed the database with initial data.
```
Comment (manio143) [inline sources/metrics/Stride.Metrics.ServerApp/Readme.md:6]: I don't see this in Program.cs

**Review by manio143 — COMMENTED** [review]
Great work! Left some suggestions.

File: sources/metrics/Stride.Metrics.ServerApp/Program.cs (hunk 6 above)
Comment (Jklawreszuk) [reply sources/metrics/Stride.Metrics.ServerApp/Program.cs:12]: Haha, yes, that was intended, just for testing purposes as some queries are very time consuming. I think I will remove it completely and try to cache some queries using views or additonal table

Comment (VaclavElias) [reply sources/metrics/Stride.Metrics.ServerApp/Program.cs:12]: Regarding the caching, from **.NET 7** you can cache very easily through minimal API, or a controller through `OutputCache`, maybe you know about that.

https://learn.microsoft.com/en-us/aspnet/core/performance/caching/output?view=aspnetcore-7.0

image

**Comment by Jklawreszuk** [comment]
I am no longer working this PR, see https://github.com/stride3d/stride/pull/2261

**Pull request closed without merging**